tests/tibanna/ffcommon/test_generic_qc_utils.py

```diff
@@ -144,4 +177,292 @@ def test_filter_workflow_args_by_property(wf_arg_set_1):
+                    "pass_target": 100,
+                    "warn_target": 80,
+                    "use_as_qc_flag": True,
+                },
+                {
+                    "id": "ts",
+                    "metric": "Total Sequences [Samtools]",
+                    "operator": ">",
+                    "pass_target": 527993296,
+                    "warn_target": 527993296,
+                },
+            ],
+            "overall_quality_status_rule": "{c1} and {ts}"
+        }
+
+
+@pytest.fixture
+def valid_ruleset_2b():
+    return {
+            "qc_thresholds": [
+                {
+                    "id": "c1",
+                    "metric": "Coverage",
+                    "operator": ">=",
+                    "pass_target": 150,
+                    "warn_target": 80,
+                    "use_as_qc_flag": True,
+                },
+            ],
+            "overall_quality_status_rule": "{c1}"
+        }
+
+
+@pytest.fixture
+def valid_ruleset_2c():
+    return {
+            "qc_thresholds": [
+                {
+                    "id": "c1",
+                    "metric": "Coverage",
+                    "operator": ">=",
+                    "pass_target": 150,
+                    "warn_target": 80,
+                    "use_as_qc_flag": True,
+                },
+            ],
+            "overall_quality_status_rule": "{c1} and {ts}",  # invalid since ts not define
+        }
+
+
+
+@pytest.fixture
+def invalid_ruleset():
+    return {
+            "qc_thresholds": [
+                {
+                    "id": "c1",
+                    "metric": "coverage",
+                    "operator": ">=",
+                    "pass_target": 100,  # warn_target missing
+                },
+            ],
+            "overall_quality_status_rule": "{c1}",
+            "applies_to": ["input_1"],
+        }
+
+
+@pytest.fixture
+def qc_json():
+    return {
+        "qc_values": [
+            {
+                "key": "Total Sequences [Samtools]",
+                "tooltip": "Total number of reads in a file, excluding supplementary and secondary reads",
+                "value": 537993296,
+            },
+            {
+                "key": "Coverage",
+                "tooltip": "Average coverage",
+                "value": 130.5,
+            },
+            {
+                "key": "test0",
+                "value": 10,
+            },
+        ],
+    }
+
+
+def test_validate_qc_ruleset_1a(valid_ruleset_1a):
+    validate_qc_ruleset(valid_ruleset_1a)  # Asserts that there is no error
+
+def test_validate_qc_ruleset_1b(valid_ruleset_1b):
+    validate_qc_ruleset(valid_ruleset_1b)  # Asserts that there is no error
+
+
+def test_validate_qc_ruleset_2(invalid_ruleset):
+    with pytest.raises(
+        GenericQcException, match="The supplied QC ruleset has the wrong format"
+    ):
+        validate_qc_ruleset(invalid_ruleset)
+
+
+def test_evaluate_qc_threshold(valid_ruleset_1a, qc_json):
```

Comment:
Structurally these tests are all very similar, you might consider using `pytest.mark.parametrize` to vary say the ruleset and the qc you want to test against. A helper function or two would be helpful with this ie:
```
@pytest.mark.parametrize('valid_ruleset, qc_json', [(ruleset1, qc1), (ruleset2, qc2)])
...
```

Reply:
Good point. I will keep that in mind. It seems a bit hard to do in this case, so I will leave it as is for now.